### DISCUSSION TITLE
Suggestion: replace minipairs with nvim-autopairs

### DISCUSSION THREAD

**Discussion by asaf** [LazyVim/LazyVim#2248] [Ideas]
I didn't feel comfortable with the editor at first glance because auto pairs acts unexpectedly in many cases, some of them are:

adding pairs when not necessary, e.g.: adding `"` before the letter _b_ near the `bar` keyword doubles pair
```ts
const a = ""bar
```

does not handle markdown code block right (e.g., ```) creates 4 pairs instead of 3 for closing code block.

I think this is super important editing is the fundamental core of everything, so far https://github.com/windwp/nvim-autopairs seem to do a great job.

**Comment by pkazmier** [comment]
Mini.pairs has been driving me bonkers for the exact reasons you mention above. I’m going to try nvim-autopairs now as well.

**Comment by Gaurav-Gosain** [comment]
I just ended up disabling it altogether!

```lua
vim.g.minipairs_disable = true
```

**Comment by nipuna-perera** [comment]
This behavior was bugging me as well. I ended up disabling mini.nvim and using vanilla nvim-autopairs. The experience is much better.

```lua
return {
  {
    "windwp/nvim-autopairs",
    event = "InsertEnter",
    opts = {},
  },
  {
    "echasnovski/mini.pairs",
    enabled = false,
  },
}
```

**Comment by AlexAllocated** [comment]
Throw me on the pile as well. I've also been annoyed by this behavior and constantly having to delete the erroneous pair. Will try out nvim-autopairs as others have suggested. I did go over mini.pairs to make sure there wasn't some setting we could toggle to fix this annoyance and I couldn't find anything.

**Comment by folke** [comment]
I like using `mini.pairs`, but I agree that the defaults can be improved.
I've just pushed an update that fixes some common annoyances: https://github.com/LazyVim/LazyVim/commit/bf809ad2a1f6cd8f74451feacdd79af68d201d6e

- don't auto close when next char is a word
- don't auto close when next char is closing and there are more closing brackets than open brackets on the line
- markdown code blocks

**Reply by dotfrag** [reply]
Thanks for improving this, it was quite annoying.. This is still not on par with `nvim-pairs` which also checks for previous un-closed pairs.

With this text and cursor position at the end of the word (pipe character = cursor):

```text
word "word|
```

`nvim-pairs` will result in:

```text
word "word"
```

`mini.pairs` will result in:

```text
word "word""
```

I wish I could help but I don't know enough, sorry. I think we need to count open pairs and see if they match close pairs?

**Reply by wadeduvall** [reply]
This is so god dang annoying... I ended up switching it manually.

**Reply by pkazmier** [reply]
Yes, me too. I just find myself fighting with mini.pairs a LOT. And the lack of support for triple characters is a deal breaker. While folke may have fixed markdown triple single quote, Python uses triple quotes a lot as well (both double and single). I just cannot see any python programmer living with mini.pairs.

**Reply by abeldekat** [reply]
A different perspective: I quite like `mini.pairs`...)

That said, I avoid most of the "conflicts" by using `mini.surround` and `snippets`(for example, fenced code block in markdown) whenever possible.

Recently, the duplication of `'` became annoying when adding a lifetime(`<'a`) in rust. Solution: 

```lua
    -- Added < to neigh_pattern to disable single quote pair in rust lifetimes
    -- ["'"] = { action = 'closeopen', pair = "''", neigh_pattern = '[^%a\\].', register = { cr = false } },
    ["'"] = { action = "closeopen", pair = "''", neigh_pattern = "[^%a\\<].", register = { cr = false } },
```

**Reply by wadeduvall** [reply]
@abeldekat of course, nvim-autopairs has this by default.

**Reply by storopoli** [reply]
If I needed to add all of the sane defaults of `nvim-autopairs` manually....

**Reply by porobertdev** [reply]
"**_Productivity_**"

it's super annoying for a beginner in vim like me. No empty line on `enter` between opening & closing tags (works fine with autopairs by default).

Another issue is with `cit` that wipes `>` of the opening tag even with autopairs.. (tested with astro & react components). Meanwhile with plain HTML `cit` goes into the deepest nested div. It only works properly (without taking out the `>`) with Astro/React components WHEN they're all on the same line (see screenshot).

<img width="561" height="197" alt="image" src="https://github.com/user-attachments/assets/1dac82a3-76ad-474d-af01-cb065544f122" />

**Comment by nipuna-perera** [comment]
Same here. It's been a while since I switched and been very happy with it.

**Comment by sqwxl** [comment]
Agree with the general sentiment here. I tried mini.pairs when I first set up lazyvim, but I quickly found it inferior to nvim-autopairs and wound up using that instead. It should be the defauld imho.

**Comment by ActionScripted** [comment]
To add to the conversation, something that's been driving me crazy with mini.pairs is that the replace window gets auto-pair'd and I always have to delete stuff in there. nvim-autopairs has `disable_in_replace_mode` and other similar options which to me make it well worth the change.

**Comment by nickjj** [comment]
After using Neovim for a few days, here's what I've seen related to mini.pairs that hasn't already been reported here. Its repo doesn't have issues enabled to report these but these are the reasons why I switched to nvim-autopairs too since it doesn't have these issues:

- In a Markdown document type \`hello , at this point if you type another \` it will insert (2) \`\` instead of just (1) \` leaving you with \`hello\`\`
  - Under normal cases if you type \` by itself it will insert \`\` so you can put your contents in between that but there's been dozens of cases where I ended up in a scenario where I had \`hello\`\` but wanted \`hello\`
- Assorted cases with ' and "" where I ended up with "hello"" instead of "hello" based on the same issue as above with Markdown
  - This feels like it could be a fundamental behavior that affects all pairs, it is repeatable

**Reply by dotfrag** [reply]
You can report issues in the parent repo https://github.com/echasnovski/mini.nvim

**Reply by nickjj** [reply]
Thanks, I reported it there in https://github.com/echasnovski/mini.nvim/issues/1731.

**Comment by jjohnston-aims** [comment]
Turning on markdown code blocks in mini.pairs was nice, but it doesn't apply to to copilot chat window.

**Comment by bulletmark** [comment]
While we are ripping into `mini.pairs` here, I'll add another one not mentioned but is infuriating. When you try to search for something containing a single quote it adds another quote in the search string. `nvim-autopairs` does not do this. Another vote to make `nvim-autopairs` the default from me!

**Comment by igmrrf** [comment]
I'm just here for the fun of it. Read every comment and I agree. I had to write scripts for many of the autopairs functionalities in order to keep my plugin list lean (use mini and snacks). Thought I could find configurations for the functionalities I want in mini.pairs but ended up writing scripts to handle them and a beautiful discussion here.
